### PR TITLE
 [14.0][MIG]Hotel: Migrated into v14

### PR DESCRIPTION
**Pull request by maitripatel19**
[14.0][MIG]Hotel: Migrated into v14

### CODE REVIEW

File: hotel/LICENSE/LICENSE [hunk 1]
```diff
@@ -0,0 +1,16 @@
+    OpenERP, Open Source Management Solution
+    Copyright (C) 2004-2010 OpenERP SA (<http://www.openerp.com>)
+    Copyright (C) 2020-TODAY Serpent Consulting Services Pvt. Ltd. (<http://www.serpentcs.com>).
```
Comment (vimalpatelserpentcs) [inline hotel/LICENSE/LICENSE:3]: `
```suggestion
    Copyright (C) 2021-TODAY Serpent Consulting Services Pvt. Ltd. (<http://www.serpentcs.com>).
```

File: hotel/models/hotel_folio.py [hunk 2]
```diff
@@ -0,0 +1,753 @@
...
+
+import time
+from datetime import datetime, timedelta
+
+from odoo import _, api, fields, models
+from odoo.exceptions import UserError, ValidationError
+from odoo.tools import DEFAULT_SERVER_DATETIME_FORMAT
+
+
+def _offset_format_timestamp1(
```
Comment (vimalpatelserpentcs) [inline hotel/models/hotel_folio.py]: No need this function, odoo DateTime class provided the timestamp method so please use that

File: hotel/models/hotel_folio.py [hunk 3]
```diff
@@ -0,0 +1,753 @@
...
+    return res
+
+
+class FolioRoomLine(models.Model):
+
+    _name = "folio.room.line"
+    _description = "Hotel Room Reservation"
+    _rec_name = "room_id"
+
+    room_id = fields.Many2one("hotel.room", "Room id")
```
Comment (vimalpatelserpentcs) [inline hotel/models/hotel_folio.py]: Add ondelete=restrict

File: hotel/models/hotel_folio.py [hunk 4]
```diff
@@ -0,0 +1,753 @@
...
+            to_zone = self._context.get("tz")
+        else:
+            to_zone = "UTC"
+        return _offset_format_timestamp1(
+            time.strftime("%Y-%m-%d 12:00:00"),
+            DEFAULT_SERVER_DATETIME_FORMAT,
+            DEFAULT_SERVER_DATETIME_FORMAT,
+            ignore_unparsable_time=True,
+            context={"tz": to_zone},
+        )
```
Comment (vimalpatelserpentcs) [inline hotel/models/hotel_folio.py]: use the Datetime class methods

File: hotel/models/hotel_folio.py [hunk 5]
```diff
@@ -0,0 +1,753 @@
...
+        "booking time or check-in "
+        "check-out time.",
+    )
+    duration = fields.Float(
+        "Duration in Days",
+        help="Number of days which will automatically "
+        "count from the check-in and check-out date. ",
+    )
+    hotel_invoice_id = fields.Many2one("account.move", "Invoice", copy=False)
+    duration_dummy = fields.Float("Duration Dummy")
```
Comment (vimalpatelserpentcs) [inline hotel/models/hotel_folio.py:114]: What is the use of duration_dummy?

File: hotel/models/hotel_folio.py [hunk 6]
```diff
@@ -0,0 +1,753 @@
...
+
+    @api.constrains("room_line_ids")
+    def folio_room_lines(self):
```
Comment (vimalpatelserpentcs) [inline hotel/models/hotel_folio.py]:
```suggestion
    def _check_duplicate_folio_room_line(self):
```

File: hotel/models/hotel_folio.py [hunk 7]
```diff
@@ -0,0 +1,753 @@
...
+                    raise ValidationError(
+                        _(
+                            """Room Duplicate Exceeded!, """
+                            """You Cannot Take Same %s Room Twice!"""
+                        )
+                        % (room_no.name)
+                    )
+
+    @api.onchange("checkout_date", "checkin_date")
+    def onchange_dates(self):
```
Comment (vimalpatelserpentcs) [inline hotel/models/hotel_folio.py]: should be a private method

File: hotel/models/hotel_folio.py [hunk 8]
```diff
@@ -0,0 +1,753 @@
...
+
+    @api.model
+    def create(self, vals):
+        """
+        Overrides orm create method.
+        @param self: The object pointer
+        @param vals: dictionary of fields value.
+        @return: new record set for hotel folio.
+        """
+        if not "service_line_ids" and "folio_id" in vals:
```
Comment (vimalpatelserpentcs) [inline hotel/models/hotel_folio.py:173]: if not "service_line_ids" and "folio_id" in vals
is it works?

File: hotel/models/hotel_folio.py [hunk 9]
```diff
@@ -0,0 +1,753 @@
...
+                for rec in folio_id:
+                    for room_rec in rec.room_line_ids:
+                        room = h_room_obj.search(
+                            [("product_id", "=", room_rec.product_id.id)]
+                        )
+                        room.write({"isroom": False})
+                        vals = {
+                            "room_id": room.id,
+                            "check_in": rec.checkin_date,
+                            "check_out": rec.checkout_date,
```
Comment (vimalpatelserpentcs) [inline hotel/models/hotel_folio.py]: Can we reduce repeated code?

File: hotel/models/hotel_folio.py [hunk 10]
```diff
@@ -0,0 +1,753 @@
...
+            if vals and vals.get("duration_dummy", False):
+                vals["duration"] = vals.get("duration_dummy", 0.0)
+            else:
+                vals["duration"] = rec.duration
+            room_lst = [folio_rec.product_id.id for folio_rec in rec.room_line_ids]
+            new_rooms = set(room_lst).difference(set(rooms_list))
+            if len(list(new_rooms)) != 0:
+                room_list = product_obj.browse(list(new_rooms))
+                for rm in room_list:
+                    room_obj = h_room_obj.search([("name", "=", rm.name)])
```
Comment (vimalpatelserpentcs) [inline hotel/models/hotel_folio.py]:
```suggestion
                    hotel_room = h_room_obj.search([("product_id", "=", rm.id)])
```

File: hotel/models/hotel_folio.py [hunk 11]
```diff
@@ -0,0 +1,753 @@
...
+        return folio_id
+
+    def write(self, vals):
+        """
+        Overrides orm write method.
+        @param self: The object pointer
+        @param vals: dictionary of fields value.
+        """
+        product_obj = self.env["product.product"]
+        h_room_obj = self.env["hotel.room"]
```
Comment (vimalpatelserpentcs) [inline hotel/models/hotel_folio.py]: Please set proper variable name
like h_room_obj -> hotel_room_obj

File: hotel/models/hotel_folio.py [hunk 12]
```diff
@@ -0,0 +1,753 @@
...
+                        "folio_id": rec.id,
+                    }
+                    folio_romline_rec = folio_room_line_obj.search(
+                        [("folio_id", "=", rec.id)]
+                    )
+                    folio_romline_rec.write(room_vals)
+        return super(HotelFolio, self).write(vals)
+
+    @api.onchange("partner_id")
+    def onchange_partner_id(self):
```
Comment (vimalpatelserpentcs) [inline hotel/models/hotel_folio.py]: please make it as private method

File: hotel/models/hotel_folio.py [hunk 13]
```diff
@@ -0,0 +1,753 @@
...
+        """
+        When you change partner_id it will update the partner_invoice_id,
+        partner_shipping_id and pricelist_id of the hotel folio as well
+        ---------------------------------------------------------------
+        @param self: object pointer
+        """
+        if self.partner_id:
+            order_ids = [folio.order_id.id for folio in self]
```
Comment (vimalpatelserpentcs) [inline hotel/models/hotel_folio.py]:
```suggestion
            order_ids = self.mapped('order_id').ids
```

File: hotel/models/hotel_folio.py [hunk 14]
```diff
@@ -0,0 +1,753 @@
...
+        """
+        if not self.order_id:
+            raise UserError(_("Order id is not available"))
+        self.invoice_ids.write({"state": "cancel"})
+        return self.order_id.action_cancel()
+
+    def action_confirm(self):
+        for order in self.order_id:
+            order.state = "sale"
+            if not order.analytic_account_id:
```
Comment (vimalpatelserpentcs) [inline hotel/models/hotel_folio.py:269]: instead of if condition we could use filtered function in for loop

File: hotel/models/hotel_folio.py [hunk 15]
```diff
@@ -0,0 +1,753 @@
...
+        config_parameter_obj = self.env["ir.config_parameter"]
+        if config_parameter_obj.sudo().get_param("sale.auto_done_setting"):
+            self.order_id.action_done()
+
+    def action_cancel_draft(self):
+        """
+        @param self: object pointer
+        """
+        if not len(self._ids):
+            return False
```
Comment (vimalpatelserpentcs) [inline hotel/models/hotel_folio.py]: No needed

File: hotel/models/hotel_folio.py [hunk 16]
```diff
@@ -0,0 +1,753 @@
...
+        query = "select id from sale_order_line \
+        where order_id IN %s and state=%s"
+        self._cr.execute(query, (tuple(self._ids), "cancel"))
+        cr1 = self._cr
+        line_ids = map(lambda x: x[0], cr1.fetchall())
```
Comment (vimalpatelserpentcs) [inline hotel/models/hotel_folio.py]: please use ORM methods instead of SQL query

File: hotel/models/hotel_folio.py [hunk 17]
```diff
@@ -0,0 +1,753 @@
...
+class HotelFolioLine(models.Model):
+
+    _name = "hotel.folio.line"
+    _description = "Hotel Folio Room Line"
+
+    @api.model
+    def _get_checkin_date(self):
+        if "checkin" in self._context:
+            return self._context["checkin"]
+        return time.strftime(DEFAULT_SERVER_DATETIME_FORMAT)
```
Comment (vimalpatelserpentcs) [inline hotel/models/hotel_folio.py]: I think no need for this function if you can pass default context on one2many fields on xml

File: hotel/models/hotel_folio.py [hunk 18]
```diff
@@ -0,0 +1,753 @@
...
+        @param vals: dictionary of fields value.
+        @return: new record set for hotel folio line.
+        """
+        if "folio_id" in vals:
+            folio = self.env["hotel.folio"].browse(vals["folio_id"])
+            vals.update({"order_id": folio.order_id.id})
+        return super(HotelFolioLine, self).create(vals)
+
+    @api.constrains("checkin_date", "checkout_date")
+    def check_dates(self):
```
Comment (vimalpatelserpentcs) [inline hotel/models/hotel_folio.py]: please make this function as private

File: hotel/models/hotel_folio.py [hunk 19]
```diff
@@ -0,0 +1,753 @@
...
+                )
+
+    def unlink(self):
+        """
+        Overrides orm unlink method.
+        @param self: The object pointer
+        @return: True/False.
+        """
+        for line in self:
+            if line.order_line_id:
```
Comment (vimalpatelserpentcs) [inline hotel/models/hotel_folio.py:358]: No need this condition

File: hotel/models/hotel_folio.py [hunk 20]
```diff
@@ -0,0 +1,753 @@
...
+        return {"domain": domain}
+
+    def button_confirm(self):
+        """
+        @param self: object pointer
+        """
+        for folio in self:
+            line = folio.order_line_id
+            line.button_confirm()
+        return True
```
Comment (vimalpatelserpentcs) [inline hotel/models/hotel_folio.py]:
```suggestion
        self.mapped('order_line_id').button_confirm()
```

File: hotel/models/hotel_folio.py [hunk 21]
```diff
@@ -0,0 +1,753 @@
...
+
+    def button_done(self):
+        """
+        @param self: object pointer
+        """
+        for rec in self:
+            lines = [folio_line.order_line_id for folio_line in rec]
+            lines.button_done()
```
Comment (vimalpatelserpentcs) [inline hotel/models/hotel_folio.py]: Please use the mapped function

File: hotel/models/hotel_room.py [hunk 22]
```diff
@@ -0,0 +1,326 @@
...
+from odoo.osv import expression
+
+
+class HotelFloor(models.Model):
+
+    _name = "hotel.floor"
+    _description = "Floor"
+
+    name = fields.Char("Floor Name", required=True, index=True)
+    sequence = fields.Integer(index=True)
```
Comment (vimalpatelserpentcs) [inline hotel/models/hotel_room.py]: index ?

File: hotel/models/hotel_room.py [hunk 23]
```diff
@@ -0,0 +1,326 @@
...
+    _description = "Hotel Room"
+
+    product_id = fields.Many2one(
+        "product.product",
+        "Product_id",
+        required=True,
+        delegate=True,
+        ondelete="cascade",
+    )
+    floor_id = fields.Many2one(
```
Comment (vimalpatelserpentcs) [inline hotel/models/hotel_room.py:30]: please add ondelete=restrict attribute

File: hotel/models/hotel_room.py [hunk 24]
```diff
@@ -0,0 +1,326 @@
...
+
+    @api.model
+    def create(self, vals):
+        if "room_categ_id" in vals:
+            room_categ = self.env["hotel.room.type"].browse(vals.get("room_categ_id"))
+            vals.update({"categ_id": room_categ.product_categ_id.id})
+        return super(HotelRoom, self).create(vals)
+
+    @api.constrains("capacity")
+    def check_capacity(self):
```
Comment (vimalpatelserpentcs) [inline hotel/models/hotel_room.py]: Please make this method as private

File: hotel/models/hotel_services.py [hunk 25]
```diff
@@ -0,0 +1,421 @@
...
+    def _service_checkin_date(self):
+        if "checkin" in self._context:
+            return self._context["checkin"]
+        return time.strftime(DEFAULT_SERVER_DATETIME_FORMAT)
+
+    @api.model
+    def _service_checkout_date(self):
+        if "checkout" in self._context:
+            return self._context["checkout"]
+        return time.strftime(DEFAULT_SERVER_DATETIME_FORMAT)
```
Comment (vimalpatelserpentcs) [inline hotel/models/hotel_services.py]: pass context on o2m fields on xml

File: hotel/models/hotel_services.py [hunk 26]
```diff
@@ -0,0 +1,421 @@
...
+    def unlink(self):
+        """
+        Overrides orm unlink method.
+        @param self: The object pointer
+        @return: True/False.
+        """
+        s_line_obj = self.env["sale.order.line"]
+        for line in self:
+            if line.service_line_id:
+                sale_unlink_obj = s_line_obj.browse([line.service_line_id.id])
```
Comment (vimalpatelserpentcs) [inline hotel/models/hotel_services.py]: No need to browse for the recordset record

File: hotel/views/hotel_room.xml [hunk 27]
```diff
@@ -0,0 +1,300 @@
+<?xml version="1.0" encoding="utf-8" ?>
```
Comment (vimalpatelserpentcs) [inline hotel/views/hotel_room.xml]: No need this line

File: hotel/views/hotel_room.xml [hunk 28]
```diff
@@ -0,0 +1,300 @@
...
+    <!--======== Floor ======== -->
+    <!-- Form view of hotel floor -->
+    <record id="view_hotel_floor_form" model="ir.ui.view">
+        <field name="name">hotel.floor.form</field>
+        <field name="model">hotel.floor</field>
+        <field name="arch" type="xml">
+            <form string=" Hotel Floor">
+                <sheet>
+                    <group>
+                        <field name="name" colspan="1" />
```
Comment (vimalpatelserpentcs) [inline hotel/views/hotel_room.xml]: colspan ?

File: hotel/models/hotel_folio.py (hunk 5 above)
Comment (maitripatel19) [reply hotel/models/hotel_folio.py:114]: I remove this field

File: hotel/models/hotel_folio.py (hunk 8 above)
Comment (maitripatel19) [reply hotel/models/hotel_folio.py:173]: yes, is it works

File: hotel/models/hotel_folio.py (hunk 19 above)
Comment (maitripatel19) [reply hotel/models/hotel_folio.py:358]: order_line_id is required for unlink.

File: hotel/models/hotel_room.py (hunk 22 above)
Comment (maitripatel19) [reply hotel/models/hotel_room.py]: it's for the most searchable fields in the model. In this way, we can speed the database search performance.
should it be removed ?

Comment (vimalpatelserpentcs) [reply hotel/models/hotel_room.py]: Yes correct, on name field would be fine but I don't think that to need the index=True on the sequence field

File: hotel/models/hotel_folio.py (hunk 9 above)
Comment (maitripatel19) [reply hotel/models/hotel_folio.py]: here, code is being repeated for try and exception,  should it do reduced by written in common method?

**Pull request merged**